### PR TITLE
Added PrefetchIterator

### PR DESCRIPTION
**Pull request by t-novak**
Useful for construct like:
```java
Dataset<T> data = Join.of(input, config)...;
ReduceByKey.of(data)
  .reduceBy((Stream<T> values, Collector<X> collector) -> {
    final PrefetchIterator<T> iterator = new PrefetchIterator<>(values);
    T first = iterator.prefetch();
    if (first != null) {
      if (first.selected())
        processSelected(iterator, collector);
      else
        processUnselected(iterator, collector);
    }
  });
```

### CODE REVIEW

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/util/PrefetchIterator.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      return it.next();
+    }
+  }
+
+  /**
+   * Get next value without moving the iterator forward.
+   *
+   * @return Next value (or null on the end)
+   */
+  public T prefetch() {
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/util/PrefetchIterator.java]: Can be name this `peek`? That is compliant with naming conventions of other methods performing this task.

Comment (t-novak) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/util/PrefetchIterator.java]: I'm in doubt about `peek()`.

It has different meaning in java 8 stream API and maybe someone would like to add `PeekingStream` later doing the same thing on stream.

Comment (je-ik) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/util/PrefetchIterator.java]: I'd say, that the naming of `Stream#peek` seems unfortunate, but OK. Can we come up with some other naming that would be clear to users? I have a strong feeling that the current naming is not self-explanatory.

Comment (t-novak) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/util/PrefetchIterator.java]: OK, more names:

- `LookupIterator::lookup()`
- `PeekingIterator::peek()`
- `PeepIterator::peep()`
- `SurveyIterator::survey()`
- `InspectIterator::inspect()`
- ?

Comment (je-ik) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/util/PrefetchIterator.java]: I still vote for `peek`, but maybe others might have some opinion on this.

Comment (je-ik) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/util/PrefetchIterator.java]: @dmvk Cool, I wasn't aware of that. I think `Iterators.peekingIterator(values.iterator())` does exactly this and we shouldn't reinvent the wheel. Also, as discussed over slack, motivation for this is probably missing #131 
We should probably close this.

**Pull request closed without merging**